docs/spec/overload.rst

```diff
@@ -183,40 +184,47 @@ Overlapping overloads
 
 If two overloads can accept the same set of arguments, they are said
 to "partially overlap". If two overloads partially overlap, the return type
-of the latter overload should be assignable to the return type of the
-former overload. If this condition doesn't hold, it is indicative of a
+of the former overload should be assignable to the return type of the
+latter overload. If this condition doesn't hold, it is indicative of a
 programming error and should be reported by type checkers::
 
   # These overloads partially overlap because both accept an
-  # argument of type ``Literal[0]``, but their return types
-  # differ.
+  # argument of type Literal[0], but the return type int is
+  # not assignable to str.
```

Comment:
I feel like I'm making too many comments on this PR, so feel free to ignore/resolve, but: would it make sense to give the rationale for some of these rules? For instance, a shortened version of my description of the rationale for the partial-overlap here: https://github.com/python/typing/pull/1839/commits/cce3879e735f401fdd4fe0654913d0d23c351076#r1733640526

I realize the rationale is not strictly required for the spec to clarify how type checkers should behave. But practically speaking, understanding the rationale can be very useful for readers of the spec, including type checker authors, and future maintainers/updaters of the spec. For the benefit of reviewers of this PR, you dug up and linked a past comment on a mypy issue that dug into the rationale for the rules specified here. Can we save future spec maintainers some similar digging, by preserving some of that rationale in the spec text itself?

(If you think this is a reasonable idea, but don't have time to do it yourself, I'm also willing to do it as a separate follow-up PR.)

Reply:
The current text justifies this by saying "If this condition doesn't hold, it is indicative of a programming error". I think it's fine to expand on this somewhat, but I don't know that we need to go into a lot more detail.

Could you remind me how you phrased your rationale? The link above doesn't seem to work.

Reply:
Here's how I phrased the rationale. I think it requires a short example for clarity:

```
The purpose of the partial-overlap rule is to prevent unsoundness of this form:

```
@overload
def is_one(x: Literal[1]) -> True: ...
@overload
def is_one(x: int) -> False: ...

def f(x: int) -> bool:
    return is_one(x)

f(1)  # Presumably True at runtime, but type checker will say False
```
```